In [6]:
# ----- Core Python & Visualization ----- #
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# ----- Data Preprocessing ----- #
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder, LabelEncoder
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline

# ----- Models ----- #
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import GradientBoostingClassifier, AdaBoostClassifier
from xgboost import XGBClassifier

# ----- Evaluation Metrics ----- #
from sklearn.metrics import r2_score, mean_squared_error, classification_report, confusion_matrix, accuracy_score

# ----- Neural Network (Keras) ----- #
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam, SGD

In [2]:
import pandas as pd 

df = pd.read_csv("/Users/anhuynh/Downloads/diabetes_binary_health_indicators_BRFSS2015.csv")

df

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0                   0.0     1.0       1.0        1.0  40.0     1.0     0.0   
1                   0.0     0.0       0.0        0.0  25.0     1.0     0.0   
2                   0.0     1.0       1.0        1.0  28.0     0.0     0.0   
3                   0.0     1.0       0.0        1.0  27.0     0.0     0.0   
4                   0.0     1.0       1.0        1.0  24.0     0.0     0.0   
...                 ...     ...       ...        ...   ...     ...     ...   
253675              0.0     1.0       1.0        1.0  45.0     0.0     0.0   
253676              1.0     1.0       1.0        1.0  18.0     0.0     0.0   
253677              0.0     0.0       0.0        1.0  28.0     0.0     0.0   
253678              0.0     1.0       0.0        1.0  23.0     0.0     0.0   
253679              1.0     1.0       1.0        1.0  25.0     0.0     0.0   

        HeartDiseaseorAttack  PhysActivity  Fruits  ...  AnyHealthcare  \
0                        0.0           0.0     0.0  ...            1.0   
1                        0.0           1.0     0.0  ...            0.0   
2                        0.0           0.0     1.0  ...            1.0   
3                        0.0           1.0     1.0  ...            1.0   
4                        0.0           1.0     1.0  ...            1.0   
...                      ...           ...     ...  ...            ...   
253675                   0.0           0.0     1.0  ...            1.0   
253676                   0.0           0.0     0.0  ...            1.0   
253677                   0.0           1.0     1.0  ...            1.0   
253678                   0.0           0.0     1.0  ...            1.0   
253679                   1.0           1.0     1.0  ...            1.0   

        NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   Age  \
0               0.0      5.0      18.0      15.0       1.0  0.0   9.0   
1               1.0      3.0       0.0       0.0       0.0  0.0   7.0   
2               1.0      5.0      30.0      30.0       1.0  0.0   9.0   
3               0.0      2.0       0.0       0.0       0.0  0.0  11.0   
4               0.0      2.0       3.0       0.0       0.0  0.0  11.0   
...             ...      ...       ...       ...       ...  ...   ...   
253675          0.0      3.0       0.0       5.0       0.0  1.0   5.0   
253676          0.0      4.0       0.0       0.0       1.0  0.0  11.0   
253677          0.0      1.0       0.0       0.0       0.0  0.0   2.0   
253678          0.0      3.0       0.0       0.0       0.0  1.0   7.0   
253679          0.0      2.0       0.0       0.0       0.0  0.0   9.0   

        Education  Income  
0             4.0     3.0  
1             6.0     1.0  
2             4.0     8.0  
3             3.0     6.0  
4             5.0     4.0  
...           ...     ...  
253675        6.0     7.0  
253676        2.0     4.0  
253677        5.0     2.0  
253678        5.0     1.0  
253679        6.0     2.0  

[253680 rows x 22 columns]

In [8]:
# Dataset Overview
print("🧠 Shape of df:", df.shape)

# Columns and Data Types
print("\n📋 Column names:\n", df.columns.tolist())
print("\n🔎 Data types:\n", df.dtypes)

# Preview
print("\n🧾 First 5 rows:\n", df.head())

# Missing Data
print("\n🧹 Missing values per column:\n", df.isnull().sum())

# Descriptive Stats for Numeric Columns
print("\n📊 Summary stats (numeric):\n", df.describe())

# Descriptive Stats for Categorical Columns
#print("\n🧩 Summary stats (categorical):\n", df.describe(include=['object', 'category']))

# Unique Values Per Column
print("\n🔢 Unique values per column:")
for col in df.columns:
    print(f"  - {col}: {df[col].nunique()} unique values")

# Duplicate Check
print("\n📎 Number of duplicate rows:", df.duplicated().sum())

# Total number of missing values
total_na = df.isna().sum().sum()
print("🧮 Total number of missing values in df:", total_na)

🧠 Shape of df: (253680, 22)

📋 Column names:
 ['Diabetes_binary', 'HighBP', 'HighChol', 'CholCheck', 'BMI', 'Smoker', 'Stroke', 'HeartDiseaseorAttack', 'PhysActivity', 'Fruits', 'Veggies', 'HvyAlcoholConsump', 'AnyHealthcare', 'NoDocbcCost', 'GenHlth', 'MentHlth', 'PhysHlth', 'DiffWalk', 'Sex', 'Age', 'Education', 'Income']

🔎 Data types:
 Diabetes_binary         float64
HighBP                  float64
HighChol                float64
CholCheck               float64
BMI                     float64
Smoker                  float64
Stroke                  float64
HeartDiseaseorAttack    float64
PhysActivity            float64
Fruits                  float64
Veggies                 float64
HvyAlcoholConsump       float64
AnyHealthcare           float64
NoDocbcCost             float64
GenHlth                 float64
MentHlth                float64
PhysHlth                float64
DiffWalk                float64
Sex                     float64
Age                     float64
Education         

In [3]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam

# Features and target
target_col = 'Diabetes_binary'
X = df.drop(columns=[target_col])
y = df[target_col]

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# Scale numeric features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


1st Basic NN model 

In [ ]:
# Build the neural network
model = Sequential([
    Dense(32, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    Dropout(0.3),
    Dense(16, activation='relu'),
    Dropout(0.2),
    Dense(1, activation='sigmoid')  # Output layer for binary classification
])

# Compile model
model.compile(optimizer=Adam(learning_rate=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train the model
history = model.fit(X_train_scaled, y_train,
                    validation_data=(X_test_scaled, y_test),
                    epochs=30,
                    batch_size=32,
                    verbose=1)




Epoch 1/30


/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 892us/step - accuracy: 0.8602 - loss: 0.3471 - val_accuracy: 0.8629 - val_loss: 0.3173
Epoch 2/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 5s 836us/step - accuracy: 0.8635 - loss: 0.3193 - val_accuracy: 0.8632 - val_loss: 0.3169
Epoch 3/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 890us/step - accuracy: 0.8642 - loss: 0.3170 - val_accuracy: 0.8627 - val_loss: 0.3166
Epoch 4/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 899us/step - accuracy: 0.8657 - loss: 0.3157 - val_accuracy: 0.8641 - val_loss: 0.3160
Epoch 5/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 12s 2ms/step - accuracy: 0.8632 - loss: 0.3175 - val_accuracy: 0.8620 - val_loss: 0.3173
Epoch 6/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 13s 2ms/step - accuracy: 0.8663 - loss: 0.3139 - val_accuracy: 0.8650 - val_loss: 0.3157
Epoch 7/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 8s 1ms/step - accuracy: 0.8650 - loss: 0.3157 - val_accuracy: 0.8630 - val_loss: 0.3168
Epoch 8/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 7s 1ms/step - accuracy: 0.8672 - loss: 0.3111 - v

In [6]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Get predicted probabilities and convert to binary predictions
y_pred_probs = model.predict(X_test_scaled)
y_pred = (y_pred_probs > 0.5).astype(int)

# Evaluate
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Display results
print(f"Accuracy : {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall   : {recall:.4f}")
print(f"F1 Score : {f1:.4f}")


1586/1586 ━━━━━━━━━━━━━━━━━━━━ 1s 453us/step
Accuracy : 0.8645
Precision: 0.6012
Recall   : 0.0819
F1 Score : 0.1442


This basic neural network model performed well in terms of accuracy but not F1 score, which can be misleading given the urgency and sensitivity of medical diagnoses. It’s important to explore additional models to better align with the dataset’s intended use to carefully and thoughtfully support accurate patient diagnoses.

2nd More in depth NN model 

In [36]:
# Build the neural network
model = Sequential([
    Dense(128, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    Dropout(0.3),
    Dense(64, activation='relu'),
    Dropout(0.2),
    Dense(1, activation='sigmoid')  # Output layer for binary classification
])

# Compile model
model.compile(optimizer=Adam(learning_rate=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train the model
history = model.fit(X_train_scaled, y_train,
                    validation_data=(X_test_scaled, y_test),
                    epochs=30,
                    batch_size=32,
                    verbose=1)

# Evaluate
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Display results
print(f"Accuracy : {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall   : {recall:.4f}")
print(f"F1 Score : {f1:.4f}")

Epoch 1/30


/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


6342/6342 ━━━━━━━━━━━━━━━━━━━━ 8s 1ms/step - accuracy: 0.8625 - loss: 0.3282 - val_accuracy: 0.8646 - val_loss: 0.3162
Epoch 2/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 1ms/step - accuracy: 0.8658 - loss: 0.3154 - val_accuracy: 0.8632 - val_loss: 0.3174
Epoch 3/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 968us/step - accuracy: 0.8657 - loss: 0.3154 - val_accuracy: 0.8660 - val_loss: 0.3150
Epoch 4/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 956us/step - accuracy: 0.8647 - loss: 0.3157 - val_accuracy: 0.8639 - val_loss: 0.3165
Epoch 5/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 956us/step - accuracy: 0.8671 - loss: 0.3129 - val_accuracy: 0.8656 - val_loss: 0.3157
Epoch 6/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 969us/step - accuracy: 0.8669 - loss: 0.3136 - val_accuracy: 0.8646 - val_loss: 0.3154
Epoch 7/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 7s 1ms/step - accuracy: 0.8672 - loss: 0.3141 - val_accuracy: 0.8653 - val_loss: 0.3152
Epoch 8/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 1ms/step - accuracy: 0.8671 - loss: 0.3127 - val

NN model with class wieights

In [7]:
from sklearn.utils import class_weight
import numpy as np

# Compute class weights
weights = class_weight.compute_class_weight(
    class_weight='balanced',
    classes=np.unique(y_train),
    y=y_train
)

class_weights = dict(enumerate(weights))

# Train with class weights
history = model.fit(X_train_scaled, y_train,
                    validation_data=(X_test_scaled, y_test),
                    epochs=30,
                    batch_size=32,
                    verbose=1,
                    class_weight=class_weights)

from sklearn.metrics import f1_score

# After training
y_pred_probs = model.predict(X_test_scaled)
y_pred = (y_pred_probs > 0.5).astype(int)

print("F1 Score:", f1_score(y_test, y_pred))



Epoch 1/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 7s 1ms/step - accuracy: 0.7683 - loss: 0.5297 - val_accuracy: 0.7228 - val_loss: 0.4898
Epoch 2/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 7s 1ms/step - accuracy: 0.7222 - loss: 0.5098 - val_accuracy: 0.7281 - val_loss: 0.4775
Epoch 3/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 866us/step - accuracy: 0.7221 - loss: 0.5078 - val_accuracy: 0.7232 - val_loss: 0.4939
Epoch 4/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 868us/step - accuracy: 0.7201 - loss: 0.5060 - val_accuracy: 0.7153 - val_loss: 0.4969
Epoch 5/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 6s 922us/step - accuracy: 0.7203 - loss: 0.5097 - val_accuracy: 0.7279 - val_loss: 0.4750
Epoch 6/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 5s 856us/step - accuracy: 0.7198 - loss: 0.5076 - val_accuracy: 0.7297 - val_loss: 0.4846
Epoch 7/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 5s 840us/step - accuracy: 0.7220 - loss: 0.5066 - val_accuracy: 0.7184 - val_loss: 0.5026
Epoch 8/30
6342/6342 ━━━━━━━━━━━━━━━━━━━━ 5s 861us/step - accuracy: 0.7199 - los

In [8]:
# Evaluate
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Display results
print(f"Accuracy : {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall   : {recall:.4f}")
print(f"F1 Score : {f1:.4f}")


Accuracy : 0.7224
Precision: 0.3074
Recall   : 0.7920
F1 Score : 0.4429


This was the best-performing model in terms of both accuracy and F1 score. Its strong performance may be attributed to its deeper architecture, with 128 and 64 units in the hidden layers, providing sufficient capacity to learn complex patterns within the data.


NN model with Random oversampler - an alternative to smote 

In [33]:
from imblearn.over_sampling import RandomOverSampler

# --- Apply Random OverSampler (ROS) ---
ros = RandomOverSampler(random_state=42)
X_train_resampled, y_train_resampled = ros.fit_resample(X_train, y_train)

# ---  Build Neural Network ---
model = Sequential([
    Dense(32, activation='relu', input_shape=(X_train_resampled.shape[1],)),
    Dropout(0.3),
    Dense(16, activation='relu'),
    Dropout(0.2),
    Dense(1, activation='sigmoid')  # Binary output
])

model.compile(optimizer=Adam(learning_rate=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])

# ---  Train Model ---
history = model.fit(X_train_resampled, y_train_resampled,
                    validation_data=(X_test, y_test),
                    epochs=30,
                    batch_size=32,
                    verbose=1)

# ---  Evaluate on Test Set ---
y_pred_probs = model.predict(X_test)
y_pred = (y_pred_probs > 0.5).astype(int)

# ---  Print Evaluation Metrics ---
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))

print("\nClassification Report:\n", classification_report(y_test, y_pred, digits=3))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Epoch 1/30


/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


10917/10917 ━━━━━━━━━━━━━━━━━━━━ 10s 842us/step - accuracy: 0.7304 - loss: 0.5368 - val_accuracy: 0.6955 - val_loss: 0.5125
Epoch 2/30
10917/10917 ━━━━━━━━━━━━━━━━━━━━ 9s 857us/step - accuracy: 0.7510 - loss: 0.5084 - val_accuracy: 0.6939 - val_loss: 0.5275
Epoch 3/30
10917/10917 ━━━━━━━━━━━━━━━━━━━━ 12s 1ms/step - accuracy: 0.7516 - loss: 0.5083 - val_accuracy: 0.7087 - val_loss: 0.5045
Epoch 4/30
10917/10917 ━━━━━━━━━━━━━━━━━━━━ 9s 820us/step - accuracy: 0.7525 - loss: 0.5065 - val_accuracy: 0.7085 - val_loss: 0.5060
Epoch 5/30
10917/10917 ━━━━━━━━━━━━━━━━━━━━ 9s 840us/step - accuracy: 0.7507 - loss: 0.5074 - val_accuracy: 0.7005 - val_loss: 0.5241
Epoch 6/30
10917/10917 ━━━━━━━━━━━━━━━━━━━━ 9s 817us/step - accuracy: 0.7514 - loss: 0.5072 - val_accuracy: 0.6949 - val_loss: 0.5198
Epoch 7/30
10917/10917 ━━━━━━━━━━━━━━━━━━━━ 9s 831us/step - accuracy: 0.7514 - loss: 0.5061 - val_accuracy: 0.7046 - val_loss: 0.5295
Epoch 8/30
10917/10917 ━━━━━━━━━━━━━━━━━━━━ 9s 825us/step - accuracy: 0.7

The neural network model with RandomOverSampler (ROS) achieved the second highest F1 score among all models, making it the best overall performer. This is particularly important in healthcare, where recall is critical — missing a diabetic patient (a false negative) can have serious consequences. At the same time, maintaining precision is also essential to avoid unnecessary tests and stress caused by false positives.



XGB classification model 

In [35]:
from xgboost import XGBClassifier


# ---  Train XGBoost Classifier ---
xgb_model = XGBClassifier(
    n_estimators=100,
    max_depth=4,
    learning_rate=0.1,
    use_label_encoder=False,
    eval_metric='logloss',
    random_state=42
)

xgb_model.fit(X_train_resampled, y_train_resampled)

# ---  Predict and Evaluate ---
y_pred = xgb_model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))

print("\nClassification Report:\n", classification_report(y_test, y_pred, digits=3))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))


/opt/anaconda3/lib/python3.12/site-packages/xgboost/core.py:158: UserWarning: [20:40:12] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


Accuracy: 0.7210659098076316
Precision: 0.3062742738362234
Recall: 0.7920497948790494
F1 Score: 0.4417357001972387

Classification Report:
               precision    recall  f1-score   support

           0      0.955     0.710     0.814     43667
           1      0.306     0.792     0.442      7069

    accuracy                          0.721     50736
   macro avg      0.630     0.751     0.628     50736
weighted avg      0.864     0.721     0.762     50736

Confusion Matrix:
 [[30985 12682]
 [ 1470  5599]]


The XGBoost model performed reasonably well compared to the neural network. However, its F1 score and accuracy were slightly lower, making it a less optimal performer overall.

Balanced Dataset - Basic NN model

In [38]:
import pandas as pd 

df_balanced = pd.read_csv("/Users/anhuynh/Downloads/diabetes_binary_5050split_health_indicators_BRFSS2015.csv")

df_balanced

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0                  0.0     1.0       0.0        1.0  26.0     0.0     0.0   
1                  0.0     1.0       1.0        1.0  26.0     1.0     1.0   
2                  0.0     0.0       0.0        1.0  26.0     0.0     0.0   
3                  0.0     1.0       1.0        1.0  28.0     1.0     0.0   
4                  0.0     0.0       0.0        1.0  29.0     1.0     0.0   
...                ...     ...       ...        ...   ...     ...     ...   
70687              1.0     0.0       1.0        1.0  37.0     0.0     0.0   
70688              1.0     0.0       1.0        1.0  29.0     1.0     0.0   
70689              1.0     1.0       1.0        1.0  25.0     0.0     0.0   
70690              1.0     1.0       1.0        1.0  18.0     0.0     0.0   
70691              1.0     1.0       1.0        1.0  25.0     0.0     0.0   

       HeartDiseaseorAttack  PhysActivity  Fruits  ...  AnyHealthcare  \
0                       0.0           1.0     0.0  ...            1.0   
1                       0.0           0.0     1.0  ...            1.0   
2                       0.0           1.0     1.0  ...            1.0   
3                       0.0           1.0     1.0  ...            1.0   
4                       0.0           1.0     1.0  ...            1.0   
...                     ...           ...     ...  ...            ...   
70687                   0.0           0.0     0.0  ...            1.0   
70688                   1.0           0.0     1.0  ...            1.0   
70689                   1.0           0.0     1.0  ...            1.0   
70690                   0.0           0.0     0.0  ...            1.0   
70691                   1.0           1.0     1.0  ...            1.0   

       NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   Age  \
0              0.0      3.0       5.0      30.0       0.0  1.0   4.0   
1              0.0      3.0       0.0       0.0       0.0  1.0  12.0   
2              0.0      1.0       0.0      10.0       0.0  1.0  13.0   
3              0.0      3.0       0.0       3.0       0.0  1.0  11.0   
4              0.0      2.0       0.0       0.0       0.0  0.0   8.0   
...            ...      ...       ...       ...       ...  ...   ...   
70687          0.0      4.0       0.0       0.0       0.0  0.0   6.0   
70688          0.0      2.0       0.0       0.0       1.0  1.0  10.0   
70689          0.0      5.0      15.0       0.0       1.0  0.0  13.0   
70690          0.0      4.0       0.0       0.0       1.0  0.0  11.0   
70691          0.0      2.0       0.0       0.0       0.0  0.0   9.0   

       Education  Income  
0            6.0     8.0  
1            6.0     8.0  
2            6.0     8.0  
3            6.0     8.0  
4            5.0     8.0  
...          ...     ...  
70687        4.0     1.0  
70688        3.0     6.0  
70689        6.0     4.0  
70690        2.0     4.0  
70691        6.0     2.0  

[70692 rows x 22 columns]

In [39]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam

# Features and target
target_col = 'Diabetes_binary'
X = df_balanced.drop(columns=[target_col])
y = df_balanced[target_col]

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# Scale numeric features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
# Build the neural network
model = Sequential([
    Dense(128, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    Dropout(0.3),
    Dense(64, activation='relu'),
    Dropout(0.2),
    Dense(1, activation='sigmoid')  # Output layer for binary classification
])

# Compile model
model.compile(optimizer=Adam(learning_rate=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train the model
history = model.fit(X_train_scaled, y_train,
                    validation_data=(X_test_scaled, y_test),
                    epochs=30,
                    batch_size=32,
                    verbose=1)



Epoch 1/30


/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1768/1768 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - accuracy: 0.7343 - loss: 0.5335 - val_accuracy: 0.7487 - val_loss: 0.5047
Epoch 2/30
1768/1768 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7482 - loss: 0.5132 - val_accuracy: 0.7519 - val_loss: 0.5059
Epoch 3/30
1768/1768 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7517 - loss: 0.5099 - val_accuracy: 0.7517 - val_loss: 0.5020
Epoch 4/30
1768/1768 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7541 - loss: 0.5064 - val_accuracy: 0.7508 - val_loss: 0.5014
Epoch 5/30
1768/1768 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7575 - loss: 0.5020 - val_accuracy: 0.7516 - val_loss: 0.5019
Epoch 6/30
1768/1768 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7568 - loss: 0.5005 - val_accuracy: 0.7528 - val_loss: 0.5015
Epoch 7/30
1768/1768 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7550 - loss: 0.5051 - val_accuracy: 0.7511 - val_loss: 0.5022
Epoch 8/30
1768/1768 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7546 - loss: 0.5027 - val_accurac

ValueError: Found input variables with inconsistent numbers of samples: [14139, 50736]

In [41]:
# ---  Evaluate on Test Set ---
y_pred_probs = model.predict(X_test)
y_pred = (y_pred_probs > 0.5).astype(int)

# Evaluate
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Display results
print(f"Accuracy : {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall   : {recall:.4f}")
print(f"F1 Score : {f1:.4f}")

442/442 ━━━━━━━━━━━━━━━━━━━━ 0s 620us/step
Accuracy : 0.5007
Precision: 0.5003
Recall   : 1.0000
F1 Score : 0.6669


With a balanced dataset, the F1 score becomes less critical. Although the deeper neural network performs better in terms of recall, it stil not highly effective in accurately diagnosing diabetic patients.

After testing two neural network architectures of varying depth, experimenting with two class balancing techniques, a new nn model with the balanced dataset, and comparing performance with an XGBoost model, I found that the best-performing model was the deep neural network with a unbalanced dataset. This model achieved an accuracy of 72.11%, a precision of 30.63%, a recall of 79.20%, and an F1 score of 44.17%. These results indicate that it was effective at identifying diabetic patients while also being mindful of medical resource constraints by balancing sensitivity with precision.